### PR TITLE
test-case: verify-pcm-list redesign

### PR DESCRIPTION
**Pull request by Bin-QA**
Patch 1:
remove 'both' from sof-dump-status to fix verify-pcm-list issue

Patch 2:
redesign test case, because older style have 'both' in TPLG
/proc/asound/pcm without 'both'
Current without 'both' in sof-tplgreader, can direct compare output
string for 'id','pcm','type'

Signed-off-by: Wu, BinX <binx.wu@intel.com>

### CODE REVIEW

**Comment by aiChaoSONG** [comment]
Error when run verify-pcm-list.sh:
`[ERR] id 0 mismatch, tplg type: playback, asound type: capture.`
can you also fix this？

**Comment by aiChaoSONG** [comment]
> > Error when run verify-pcm-list.sh:
> > `[ERR] id 0 mismatch, tplg type: playback, asound type: capture.`
> > can you also fix this？
> 
> I already update for fix this issue, can you try again?

The reason we got the issue is that the order of pipeline from aplay -l and tplgreader is not the same:
8 0 0 1 2 3 4 5 (tplgreader)
0 0 1 2 3 4 5 8 (aplay -l).
so, the `for idx in $(expr $PIPELINE_COUNT - 1))` doesn't work properly. when idx=1, we got "id=0,playback,port1" for tplgreader pcm list, and  "id=0,capture,port1" for 'aplay -l'.
@Bin-QA Let get this pr first merged, and then fix verify-pcm-list.sh. see if we can get an idea that pipeline order doesn't matter.

**Comment by aiChaoSONG** [comment]
@Bin-QA Seem the test case need some fix:
```
2020-02-25 07:00:30 UTC [INFO] Comparing capture device 8: DMIC16k
8 +  capture +  DMIC16k ----- 8 + capture + DMIC16k
2020-02-25 07:00:30 UTC [INFO] Comparing playback device 0: Port1
0 +  playback +  Port1 ----- 0 + capture + Port1    [NOT MATCH HERE]
2020-02-25 07:00:30 UTC [INFO] Comparing capture device 0: Port1
0 +  capture +  Port1 ----- 0 + capture + Port1
2020-02-25 07:00:30 UTC [INFO] Comparing capture device 1: DMIC
1 +  capture +  DMIC ----- 1 + capture + DMIC
2020-02-25 07:00:30 UTC [INFO] Comparing playback device 2: HDMI1
2 +  playback +  HDMI1 ----- 2 + playback + HDMI1
2020-02-25 07:00:30 UTC [INFO] Comparing playback device 3: HDMI2
3 +  playback +  HDMI2 ----- 3 + playback + HDMI2
2020-02-25 07:00:30 UTC [INFO] Comparing playback device 4: HDMI3
4 +  playback +  HDMI3 ----- 4 + playback + HDMI3
2020-02-25 07:00:30 UTC [INFO] Comparing playback device 5: Speakers
5 +  playback +  Speakers ----- 5 + playback + Speakers
2020-02-25 07:00:30 UTC [INFO] Test Result: PASS!
```

**Comment by xiulipan** [comment]
@Bin-QA seems to have some issue when try to parse PCM with both playback and capture.

**Comment by Bin-QA** [comment]
Already catch the problem:
```
unset ASOUND_PCM_0
declare -Ag ASOUND_PCM_0
ASOUND_PCM_0["id"]="0"
ASOUND_PCM_0["pcm"]="Port1"
ASOUND_PCM_0["type"]="playback"
ASOUND_PCM_LST[1]="id=0;pcm=Port1;type=capture"
unset ASOUND_PCM_0
declare -Ag ASOUND_PCM_0
ASOUND_PCM_0["id"]="0"
ASOUND_PCM_0["pcm"]="Port1"
ASOUND_PCM_0["type"]="capture"
ASOUND_PCM_LST[2]="id=1;pcm=DMIC;type=capture"
```
older style, we don't have the duplicate pipeline id because both,  
why tplg loop is not effect, because it go through by index,  
pcm need to match the pipeline, so it create a map for the pipeline id,  
current have the duplicate pipeline id, so I need refine the test-case to fix it

**Comment by xiulipan** [comment]
@Bin-QA Good to know the issue, please help to fix! Thanks!

**Comment by Bin-QA** [comment]
@xiulipan @aiChaoSONG already fixed

**Comment by xiulipan** [comment]
another issue found during test.
If do not pass any TPLG to the test, the test will result in 
```
./test-case/verify-pcm-list.sh
2020-02-25 08:56:15 UTC [INFO] Test Result: PASS!
```
could you also check and fix.

**Comment by xiulipan** [comment]
@Bin-QA PS: could you add some verbose log about what PCM is checked? Now the pass log seems very poor for debug:
```
./test-case/verify-pcm-list.sh -t sof-apl-nocodec.tplg
2020-02-25 08:58:25 UTC [INFO] ./test-case/verify-pcm-list.sh using /lib/firmware/intel/sof-tplg/sof-apl-nocodec.tplg as target TPLG to run the test case
2020-02-25 08:58:25 UTC [INFO] Run command: 'sof-tplgreader.py /lib/firmware/intel/sof-tplg/sof-apl-nocodec.tplg -s 0 -e' to get BASH Array
2020-02-25 08:58:25 UTC [INFO] Run command: 'sof-dump-status.py -i 0' to get BASH Array
2020-02-25 08:58:26 UTC [INFO] Test Result: PASS!
```

**Comment by Bin-QA** [comment]
@xiulipan already update

File: test-case/verify-pcm-list.sh [hunk 1]
```diff
@@ -69,6 +71,34 @@ func_error_process()
...
+    local type=$(func_pipeline_parse_value $1 type)
+    local pcm=$(func_pipeline_parse_value $1 pcm)
+    # filter to catch the same id & type
+    # vierfy the pcm list for keyword: type, id, pcm
+    for idx in $(seq 0 $(expr $ASOUND_PCM_COUNT - 1))
+    do
+        if [ "$id" == "$(func_asound_pcm_parse_value $idx id)" ]; then
+            # same pcm id with different pcm name it mast mismatch
+            if [ "$pcm" != "$(func_asound_pcm_parse_value $idx pcm)" ];then
+                dlogi "TPLG: '$id: $pcm' mismatch with pcm: $(func_asound_pcm_parse_value $idx pcm)"
```
Comment (xiulipan) [inline test-case/verify-pcm-list.sh]: could we also add card id here?
To make it looks more user friendly.
```
2020-02-25 10:03:21 UTC [INFO] TPLG: '0: Port0: playback' matched with pcm
```
=>
`2020-02-25 10:03:21 UTC [INFO] TPLG: 'hw:0,0: Port0: playback' matched with CARD0 PCM0P`

Comment (Bin-QA) [reply test-case/verify-pcm-list.sh]: `CARD0 PDM0P` how about this format: `Card 0: id=2;pcm=HDMI1;type=playback`

File: test-case/verify-pcm-list.sh [hunk 2]
```diff
@@ -24,6 +24,8 @@ OPT_PARM_lst['t']=1         OPT_VALUE_lst['t']="$TPLG"
 func_opt_parse_option $*
 tplg=${OPT_VALUE_lst['t']}
 
+[[ ! "$tplg" ]] && dlogw "Miss tplg file for this case" && exit 1
```
Comment (stripes416) [inline test-case/verify-pcm-list.sh]:
```suggestion
[[ ! "$tplg" ]] && dlogw "Missing tplg file for this case" && exit 1
```

File: test-case/verify-pcm-list.sh [hunk 3]
```diff
@@ -69,6 +71,34 @@ func_error_process()
...
 }
 
+func_compare_asound_pipeline_by_idx()
+{
+    local idx
+    local id=$(func_pipeline_parse_value $1 id)
+    local type=$(func_pipeline_parse_value $1 type)
+    local pcm=$(func_pipeline_parse_value $1 pcm)
+    # filter to catch the same id & type
+    # vierfy the pcm list for keyword: type, id, pcm
```
Comment (stripes416) [inline test-case/verify-pcm-list.sh]:
```suggestion
    # verify the pcm list for keyword: type, id, pcm
```

File: test-case/verify-pcm-list.sh [hunk 4]
```diff
@@ -69,6 +71,34 @@ func_error_process()
...
+    for idx in $(seq 0 $(expr $ASOUND_PCM_COUNT - 1))
+    do
+        if [ "$id" == "$(func_asound_pcm_parse_value $idx id)" ]; then
+            # same pcm id with different pcm name it mast mismatch
```
Comment (stripes416) [inline test-case/verify-pcm-list.sh]:
```suggestion
            # same pcm id with different pcm name must be mismatched
```

File: test-case/verify-pcm-list.sh [hunk 5]
```diff
@@ -69,6 +71,34 @@ func_error_process()
...
+            if [ "$pcm" != "$(func_asound_pcm_parse_value $idx pcm)" ];then
+                dlogi "TPLG: '$id: $pcm' mismatch with pcm: $(func_asound_pcm_parse_value $idx pcm)"
+                return 1
+            fi
+            # because type will with different type for the same pcm id
```
Comment (stripes416) [inline test-case/verify-pcm-list.sh]:
```suggestion
            # 'pcm name' has already been compared at this point
            # Type can match with different type for the same pcm id
```

File: tools/sof-dump-status.py [hunk 6]
```diff
@@ -136,14 +136,23 @@ def _sort_pcm(pcm_info):
 
         for line in output.splitlines():
             line_filed = line.split(':')
+            # filed 0 : "s-id" s is sound card id; id is pcm id
```
Comment (stripes416) [inline tools/sof-dump-status.py]: Do you mean field instead of filed here? This occurs quite a bit in this file.

File: tools/sof-dump-status.py [hunk 7]
```diff
@@ -282,29 +291,14 @@ def export_cardinfo_pcm(card_info):
...
             pcm = pcm_lst[idx]
-            print('%s_LST[%d]="%s;%s;%s=%s"' % (keyword, idx, _getStr(pcm, 'id'), _getStr(pcm, 'pcm'), 'type', ', '.join(pcm.get('type'))))
-            # date format for match tplg format
-            pcm['pcm'] = pcm['pcm'].replace('(*)', '').strip()
-            pcm['type'] = _covertType(pcm['type'])
-            # store pcm to each list
-            print('unset %s_%s' % (keyword, pcm['id']))
-            print('declare -Ag %s_%s' % (keyword, pcm['id']))
+            print('%s_LST[%d]="%s;%s;%s"' % (keyword, idx, _getStr(pcm, 'id'), _getStr(pcm, 'pcm'), _getStr(pcm, 'type')))
+            # date format for match tplg format store pcm to each list
```
Comment (stripes416) [inline tools/sof-dump-status.py]:
```suggestion
            # Use date format for matched tplgs when storing pcms to each list
```
If that doesn't make sense, feel free to ignore this suggestion --- I struggled a bit trying to understand what you were trying to say.

**Comment by Bin-QA** [comment]
@aiChaoSONG @xiulipan @stripes416 
Sorry please review this change, because I recheck the case logic  
I find after tplg remove 'both' can direct compare the output string, which is more direct & quickly  
the key command:
```
sof-tplgreader.py $tplg_file -d id pcm type -o
sof-dump-status.py -i ${SOFCARD:-0}
```
Change root reason:  
1. pcm without 'both', so older step is merge pcm to create 'both' then compare the content
2. tplg can sort the 'id' and custom output filed to dump the result by sof-tplgreader ( here is my mistake, I forget to use this feature )
3. check sof-tplgreader & sof-dump-status output ( sof-dump-status is this round change)

So as the format output string, they should been same

**Comment by aiChaoSONG** [comment]
@Bin-QA I prefer line by line comparison. if mismatch, we need some effort to identify which pipeline mismatched from current output. but if we do line by line comparison, we can give an output:
`TPLG:id=0;pcm=Port1;type=playback <=> ASND:id=0; pcm=Port1;type=palyback [Match]`
`TPLG:id=1;pcm=Port1;type=playback <=> ASND:id=0; pcm=Port1;type=palyback [Not Match]`
More user friendly, right?

**Comment by Bin-QA** [comment]
@aiChaoSONG 
`TPLG:id=1;pcm=Port1;type=playback <=> ASND:id=0; pcm=Port1;type=palyback [Not Match]`
this line couldn't happen, except have different 'pcm name'
TPLG line create by sof-tplgreader, which dump will use `strip()` and format output
ANSD line create by sof-dump-status, which dump also use `strip()` and format output

so this mismatch is not happen, current logic:
```
    dlogi "CMD: 'sof-tplgreader.py $tplg_file -d id pcm type -o' to get tplg list string:"
    echo "$tplg_str"
    dlogi "CMD: 'sof-dump-status.py -i ${SOFCARD:-0}' to get pcm list string:"
    echo "$pcm_str"

    if [[ "$tplg_str" != "$pcm_str" ]]; then
```
So you can see the output string already dump

compare by line with line is good idea, but it will complex,

**Comment by xiulipan** [comment]
@Bin-QA Thanks for the update.
@aiChaoSONG Yep, no need to make log that complex. The error dump is enough.

**Pull request merged**